Describe MM and MP methylation tags. (#418)
samtools/hts-specs: SAMtags.tex

  {\tt MD} & Z & String for mismatching positions \\
  {\tt MF} & ? & Reserved for backwards compatibility reasons \\
  {\tt MI} & Z & Molecular identifier; a string that uniquely identifies the molecule from which the record was derived \\
  {\tt MM} & i & Base modifications / methylation  \\

> % {\tt MP} & i & Base modification qualities \\

Reviewer: is this commented out on purpose?

Author: Yes.  MP was the previous definition using phred scale.  I could completely remove it, but kept it incase we ever want to go back to it as it's not 100% clear what people actually want!  ML though I think is more realistic, so that's what I changed to and commented out MP until we're sure it's absolutely not required.

This is then followed by a comma separated list of how many unmodified seq bases of the stated base type to skip, stored as a delta to the last and starting with 0 as the first (or next) base, starting from the uncomplemented 5' end of the {\sf SEQ} field.
This number series is comparable to the numbers in an {\tt MD} tag,
albeit counting specific base types only and potentially reverse-complemented.

For example {\tt C+m,5,12,0;} tells us there are three 5-Methylcytosine bases in the original orientation {\sf SEQ}.
The first 5 {\tt C} bases are unmodified and the 6th is modified, as are the 19th (with 12 between the 6th and 19th) and 20th.
Similarly {\tt G-m,14;} indicates the 15th {\tt G} is a 5-Methylcytosine on the opposite strand (still counting from the 5' end).

This permits modifications to be listed on either strand with the rare potential for both strands to have a modification at the same site.

> Note it is permitted for the coordinate list to be empty (for example {\tt Mm:Z:C+m;}), which may be used as an explicit indicator that this base modification is not present.

Reviewer 2: Getting down to implementing this output in the ONT Megalodon software. In the modified basecaller the standard output will include a modified base call at every relevant base of the read. I believe this would result in the `Mm` tag: `Mm:Z:C+m,0,0,...,0` (with the number of zeros corresponding to the number of `C`s in a read). Is this correct? If so, would it make sense to allow a special exception for the "all bases include a modified base call" to be listed instead without any positions values (i.e. `Mm:Z:C+m`) instead of this being the special case where no bases are identified as modified?

Reviewer 2: After a bit of further development I have changed my mind and think that the current interpretation of the "empty" MM tag is more consistent with the tag. It also seems that thresholding the probabilities to some extent will be useful and so outputting absolutely all potential modified base probabilities will likely not be a primary use case.

%
% Quality values for ambiguity codes give the likelihood that the
% modification is one of the possible codes compatible with that
% ambiguity code.  For example {\tt MM:Z:C+C,10; MP:Z:+} indicates a C
% call with an unspecified modification and the phred score of 10 (ASCII
% value {\tt +}).  This corresponds to a 90\% chance of the base being
% modified.
%
% To represent several possible modifications at the same site the {\tt MP} tag can be used to indicate the probabilities of each possibility.
% The values used should be absolute probabilities, not relative between the alternatives.
% For example, a C base that has 95\% chance of being modified with 5mC being three times more likely than 5hmC will encode 5mC with 67.5\% probability ($0.9 * 0.75$ giving phred score 5, ASCII value {\tt \&})and 5hmC with 22.5\% probability ($0.9 * 0.25$ giving phred score 1, ASCII value {\tt "}).
% This could be represented with ``{\tt MM:Z:C+m,10;C+h,10; MP:Z:" \&}''.

\item[Ml:B:C,\tagvalue{scaled-probabilities}]
\hfill\\
The optional {\tt Ml} tag lists the probability of each modification listed in the {\tt Mm} tag being correct, in the order that they occur.

> The probabilities are scaled to fit in one byte such that $p=0.0$ is value 0 and $p=1.0$ is value 255.

Reviewer 2: Very minor point, but for the scaled probabilities it seems logical that each integer value represents an equal proportion of the input space (0-1). Thus the value 0 would represent probabilities in the range `0/256` to `1/256` and `255` value would represent the range `255/256` to `256/256==1.0`. Is this worth specifying explicitly?

Author: So with that in mind, and your comment below which I now understand, the integer value is centred on the bin.  Thus 0 on average (assuming uniform distribution, which obviously isn't likely to be true but let's let that slide) will mean p=0.5/256 and 255 means p=255.5/256?

To figure out if this is consistent, I thought about the phred scale definition where Q20 means 0.01 chance of an error (p=0.99).  If we compute an actual p=0.9895 giving phred Q19.79, then we'd be expected to round that to Q20 not Q19.  Thus similarly we're saying although Q20 is 0.01 chance of error, actually anywhere from P(err)=0.0090 to 0.0112 also give Q20.  I think the difference in language is we define phred scales in encoding (probability to integer) whereas the above definition is defined in the reverse (integer to probability).  We never explicitly state though with phred which way rounding is meant to go, and I expect some do floor while others do round (nearest).

Ie after some mental gymnastics, I agree with your assessment.  I do wonder if it's the right way to approach it though.

An alternative would be to define the integer value as being `p*256`, rounding to nearest integer and with an explicit upper limit of 255.  I think that's better than p*255 to avoid the overflow.

Reviewer 2: I'm happy to leave this unspecified, but it seems to my mind that the most logical thing to do would be to have each integer encoding represent an equal range of the probability space. By not including this in the specification, alternative interpretations are likely to result. The issue with the `p*256` rounded and then explicitly clipping to `255` is that the value `0` represents `1/512` stretch of input probability, while `255` represents `3/512` stretch of the input probability space, while all other encodings represent `1/265`.

Reviewer 2: I do think my formula below is equivalent to `p*256` then floor then clip to `255`.

Author: The only problem with `p*256` instead of `p*255` is it removes the possibility of stating a certainty.  Ie this base *is* modified.  However I don't see that as a problem given the current (and likely future) technologies.

A value of 255 still corresponds to p=0.996 or Q24 in Phred Scale.  If we ever get technologies able to cope assign modification likelihoods higher than that for individual reads (not consensii) then we can introduce an MP tag to use a different scale, as originally proposed in this PR.

I'll change this PR to use p*256.

\section{Draft tags}

These are tags which have been proposed and are broadly accepted to
become standard tags, but a review or probationary period has been
deemed useful.  They use the locally-defined tag namespace and
processing software should consider that the tags may have local usage
for other purposes.

\begin{center}\small
% This table is sorted alphabetically
\begin{longtable}{ccp{12.5cm}}
  \hline
  {\bf Tag} & {\bf Type} & {\bf Description} \\
  \hline
  \endhead
  {\tt Mm} & i & Base modifications / methylation  \\
% {\tt MP} & i & Base modification qualities \\
  {\tt Ml} & i & Base modification probabilities \\
\end{longtable}
\end{center}

\subsection{Base modifications}

Base modifications, including base methylation, are represented as a series of edits from the primary unmodified sequence as originally reported by the sequencing instrument.
This potentially differs to the sequence stored in the main SAM {\sf SEQ} field if the latter has been reverse complemented, in which case SAM {\sf FLAG} 0x10 must be set.
This means modification positions are also recorded against the original orientation (i.e. starting at the 5' end), and count the original base types.

> Each modified base listed also has a quality value associated with it.

Reviewer: since we are not sure that phred is appropriate for the values we will be seeing, the part that discusses quality should be commented out here too, right?

Author: Hmm interesting semantics here.  Quality value to me doesn't imply Phred.  It simply implies a numeric value indicating the quality, which may or may not be calibrated (let's face it many often aren't!) and may or may not imply phred scale.

However we should look for a better term.  Maybe "likelihood" instead of "quality value"?  That tends to mean something very specific to some stats people.  Maybe "probability associated with it" instead then?

Given the unmodified base already has a phred likelihood, this base modification quality should be interpreted as the likelihood of this modification being correct given an assumption the original call is correct.

\begin{description}
\item[Mm:Z:\tagregex{([ACGTUN][-+][a-z]+(,[0-9]+)*;)*}]
\hfill\\
The first character is the unmodified ``fundamental'' base as reported

> by the sequencing instrument.

Reviewer: what if SEQ has IUPAC codes? how does the counting work then?

Author: They'd be covered as N as it explicitly said "N for anything else".  I'll make that more detailed with an explicit clause to cover IUPAC codes.

Author: Note some people use the phrase "canonical base" instead of "fundamental base".  I'm happy with either - votes either way welcomed.

It must be one of {\tt A}, {\tt C}, {\tt G}, {\tt T}, {\tt U} (if RNA) or {\tt N}.
Note {\tt N} may be used to match any base rather than specifically an {\tt N} call by the sequencing instrument.
This may be used in situations where the base modification is not a derivation of a standard base type.

> This is followed by either plus or minus indicating the strand the modification was observed on (relative to the original sequenced strand of {\sf SEQ} with plus meaning same orientation\footnote{Hence a tool that may reverse complement sequences does not need to understand how to manipulate the {\tt Mm} and {\tt Ml} tags.}), and one or more base modification symbols.

Reviewer: We need a name for this plus or minus. so that we can refer to it. 
How about ModifiedStrand (`+` means same as SEQ and `-` means the other Strand)?

Reviewer: You have `symbol` here and `Code` below in the table.

Author: I'll change it to `code`.

As for `ModifiedStrand`, I could do something like

> This is followed by the ModifiedStrand symbol, either plus or minus indicating the strand the modification was observed on  ...

However I'm not sure where we ever refer to it elsewhere other than in literal examples which obviously use `+` and `-`.  Can you give an example where you think we need to refer back to it by that name?

This is then followed by a comma separated list of how many unmodified seq bases of the stated base type to skip, stored as a delta to the last and starting with 0 as the first (or next) base, starting from the uncomplemented 5' end of the {\sf SEQ} field.
This number series is comparable to the numbers in an {\tt MD} tag,
albeit counting specific base types only and potentially reverse-complemented.

For example {\tt C+m,5,12,0;} tells us there are three 5-Methylcytosine bases in the original orientation {\sf SEQ}.
The first 5 {\tt C} bases are unmodified and the 6th is modified, as are the 19th (with 12 between the 6th and 19th) and 20th.

> Similarly {\tt G-m,14;} indicates the 15th {\tt G} is a 5-Methylcytosine on the opposite strand (still counting from the 5' end).

Reviewer: When noting a modification of the non-sequenced strand, do we still reference the base that was originally sequenced, or the rev-comp?

Reviewer: I think that it would be good to have explicit examples for all 4 cases of nagativeStrand (in FLAG) and "ModifiedStrand" with bases so that it's clear which base is being modified...there are too many options for confusion.

Reviewer: it would be good to state here that the abbreviation should _not_ be used.

Author: It was pointed out (maybe by you?) early on that the bases must be in the original orientation as reverse complementing a sequence eg via (re)alignment may not know about these tags.  So that bit hopefully is clear.   This question however is on the non-sequenced strand rather than the orientation (REVERSE sam flag).

It does need to be explicit, and yes it's confusing as I can't remember how it works myself and had to check!  The test/SAMtags/MM-orient.sam file has examples of all 4 combinations.  Do we want concrete examples in the spec too, or just wording to clarify this?

Note "non-sequenced strand" isn't perhaps the correct wording.  We cannot record modifications on the bottom strand unless it was sequenced.  The theory here is some technologies can do double stranded sequencing (eg the 1D^2 ONT which did top strand followed by bottom strand).  Single stranded sequencing is always going to report the same strand I'd guess, and reverse complement as appropriate.  I'll make some modifications to clarify this as I see I'm wrongly saying "original orientation" when I mean "top strand".

Author: > it would be good to state here that the abbreviation should _not_ be used.

I don't understand this comment, sorry.

Author: I am wondering though about the wisdom of my reporting using the fundamental (canonical) base on the top strand.  For example:

```
       m
AGCTTAGCGGCGATCGAG
||||||||||||||||||
TCGAATCGCCGCTAGCTC
               m
```

We'd report that as C+m and G-m as it's C and G on top strand.  It may be much clearer to say C-m for the bottom case as it's a C->m transition not G->m and then we're consistent.  I'm struggling to remember why I did it the way I did.  I fear it was for a good reason, but I cannot grasp what that would be!

Reviewer 2: I'm not writing any software to output methylation on both strands yet, so happy with either resolution.

% call with an unspecified modification and the phred score of 10 (ASCII
% value {\tt +}).  This corresponds to a 90\% chance of the base being
% modified.
%
% To represent several possible modifications at the same site the {\tt MP} tag can be used to indicate the probabilities of each possibility.
% The values used should be absolute probabilities, not relative between the alternatives.
% For example, a C base that has 95\% chance of being modified with 5mC being three times more likely than 5hmC will encode 5mC with 67.5\% probability ($0.9 * 0.75$ giving phred score 5, ASCII value {\tt \&})and 5hmC with 22.5\% probability ($0.9 * 0.25$ giving phred score 1, ASCII value {\tt "}).
% This could be represented with ``{\tt MM:Z:C+m,10;C+h,10; MP:Z:" \&}''.

\item[Ml:B:C,\tagvalue{scaled-probabilities}]
\hfill\\
The optional {\tt Ml} tag lists the probability of each modification listed in the {\tt Mm} tag being correct, in the order that they occur.
The continuous probability range 0.0 to 1.0 is remapped in equal
sized portions to the discrete integers 0 to 255 inclusively. Thus the
probability range corresponding to integer value $N$ is $N/256$ to

> $(N+1)/256$.

Reviewer: so, we can say that a modification happens at 100% but we cannot state that a modification definitely did _not_ occur? it might be useful to be able to express a certainty that a modification did not occur, _n'est pas_?

Reviewer: I suggest we make this $p=N/255$ instead.

Reviewer 2: This definition defines N -> p. What is the corresponding formula for of p -> N?

Reviewer 2: I brought this up in a [thread earlier](https://github.com/samtools/hts-specs/pull/418#pullrequestreview-467822286). It seems the most logical thing to do would be to bin the space from 0 to 1 in equal sized bins. Thus we can't represent exactly 0.0 or 1.0, but instead represent the range 0-1/256 and 255/256-1. 

It seems the only other logical conversion would be to have special cases for 0.0 and 1.o probabilities, and then have the rest of the range equally spaced. But as noted by @jkbonfield the need for absolute certainty seems limited.

Author: > 
> 
> so, we can say that a modification happens at 100% but we cannot state that a modification definitely did _not_ occur? it might be useful to be able to express a certainty that a modification did not occur, _n'est pas_?

Actually we can specify that.  We list the modification code and give it $p=0$ to state the likelihood of this C being m is zero.  Infact this is basically what ONT already do, albeit possibly with a low cutoff to avoid excessive bloat.

Are you suggesting instead we take the opposite approach of eg `C+C,0,0,3,0,1;` to state that these Cs are real C and not modified ones?  I'm not sure I see the merit in that.

Reviewer: I think I was confused about how the formula for N(p) is defined. I think we should write it explicitly: 
```
N=floor(p*255) 
```

Reviewer 2: I agree adding the formula would help with clarity. It might be good to catch for the edge case where `p=1.0`. E.g. `N=min(255, floor(p*256))`

The SAM encoding therefore uses a byte array of type `{\tt C}' with the number of elements matching the summation of the number of modifications listed as being present in the {\tt Mm} tag accounting for multi-modifications each having their own probability.

For example ``{\tt Mm:Z:C+m,5,12;C+h,5,12;}'' may have an associated tag of ``{\tt Ml:B:C,204,89,26,130}''.

If the above is rewritten in the multiple-modification form, the probabilities are interleaved in the order presented, giving ``{\tt Mm:Z:C+mh,5,12;  Ml:B:C,204,26,89,130}''.

> Note where several possibile modifications are presented at the same site, the {\tt Ml} values represent the absolute probabilities of the modification call being correct and not the relative likelihood between the alternatives.

Reviewer: possibile->possible

Reviewer: if the Ml is _likelihood_ then we shouldn't confuse further by saying that it's the "absolute _probability_"....what it should be is the likelihood of making the measurement that was made assuming the modification is as stated. i.e P(data | modification) not P(modification|data).

Reviewer: to get  to a posterior probablity one would need a prior...

Author: I was fearful of using a phrase like likelihood as it can be interpreted either in a very specific statistical way or in a more casual English speaking way.  I used `l` in the tag name because I'd previously used `p` for the phred score and `MQ` tag already exists.  Maybe that's wrong and I should have gone with `MP` for this linear scale and if we ever switch use `MZ`, as we sometimes use Z already; eg barcode phred scores.

I admit my stats is pretty ropey though so I'm not really sure exaclty what it should be.  Maybe @marcus1487 can advise?  My instinct though is that for phred base calls we're describing the probability of the observed call being the real one, so that's P(observed | real)?  I think that agrees with your P(data | modification) statement.

As for priors, if I recall that's exactly what I was doing in the `samtools consensus` PR using a prior likelihood of modifications, but I'd have to do some hunting to see if that was a plan or whether I put it into fruition. :)

> These need not sum to 256 ($p \approx 1.0$) as the remainder represents the probability that none of the modification types are present.

Reviewer: this comment is confusing since the formula for getting the probabilty adds one...and so they wouldn't add to 256 even if there were no remainder....

Author: I'm confused by what you mean here.

Are you saying they wouldn't add to 1 because we're quantising to a discrete set of probabilities and not using a continuous distribution?  That's obviously true - it's just rounding error.  However this is something over and above that.
It's simply saying that if you claim it's p=0.2 of 5hmC and p=0.5 of 5mC then that's only p=0.7 of it being modified, with the remaining p=0.3 of no modification at all.

In the limit, a single modification code always does this obviously as we always have a remainder.

Reviewer: I got confused about how p is related to N....safe to ignore this.

Color read quality on the original strand of the read. Same encoding as {\sf QUAL}; same length as {\tt CS}.
\end{description}

\section{Draft tags}

These are tags which have been proposed and are broadly accepted to
become standard tags, but a review or probationary period has been
deemed useful.  They use the locally-defined tag namespace and
processing software should consider that the tags may have local usage
for other purposes.

\begin{center}\small
% This table is sorted alphabetically
\begin{longtable}{ccp{12.5cm}}
  \hline
  {\bf Tag} & {\bf Type} & {\bf Description} \\
  \hline
  \endhead

>   {\tt Mm} & i & Base modifications / methylation  \\
> % {\tt MP} & i & Base modification qualities \\
>   {\tt Ml} & i & Base modification probabilities \\

Reviewer 3: ```suggestion
  {\tt Mm} & Z & Base modifications / methylation  \\
% {\tt MP} & Z & Base modification qualities \\
  {\tt Ml} & B,C & Base modification probabilities \\
```

\end{longtable}
\end{center}

\subsection{Base modifications}

Base modifications, including base methylation, are represented as a series of edits from the primary unmodified sequence as originally reported by the sequencing instrument.
This potentially differs to the sequence stored in the main SAM {\sf SEQ} field if the latter has been reverse complemented, in which case SAM {\sf FLAG} 0x10 must be set.
This means modification positions are also recorded against the original orientation (i.e. starting at the 5' end), and count the original base types.

Each modified base listed also has a quality value associated with it.
Given the unmodified base already has a phred likelihood, this base modification quality should be interpreted as the likelihood of this modification being correct given an assumption the original call is correct.

\begin{description}

> \item[Mm:Z:\tagregex{([ACGTUN][-+][a-z]+(,[0-9]+)*;)*}]

Reviewer 3: Needs to admit the ChEBI notation, something like:

```suggestion
\item[Mm:Z:\tagregex{([ACGTUN][-+]([a-z]+|[0-9]+)(,[0-9]+)*;)*}]
```

\hfill\\
The first character is the unmodified ``fundamental'' base as reported
by the sequencing instrument for the top strand.
It must be one of {\tt A}, {\tt C}, {\tt G}, {\tt T}, {\tt U} (if RNA) or {\tt N} for anything else, including any IUPAC ambiguity codes in the reported SEQ field.
Note {\tt N} may be used to match any base rather than specifically an {\tt N} call by the sequencing instrument.
This may be used in situations where the base modification is not a derivation of a standard base type.

> This is followed by either plus or minus indicating the strand the modification was observed on (relative to the original sequenced strand of {\sf SEQ} with plus meaning same orientation\footnote{Hence a tool that may reverse complement sequences does not need to understand how to manipulate the {\tt Mm} and {\tt Ml} tags.}), and one or more base modification codes.

Reviewer 3: Minor editorial — footnotes in these documents follow the convention of going ideally at the end of the sentence and always after the punctuation:
```suggestion
This is followed by either plus or minus indicating the strand the modification was observed on (relative to the original sequenced strand of {\sf SEQ} with plus meaning same orientation),\footnote{Hence a tool that may reverse complement sequences does not need to understand how to manipulate the {\tt Mm} and {\tt Ml} tags.} and one or more base modification codes.
```

It is not permitted for coordinates to be beyond the length of the sequence.

When multiple modifications are listed, for example {\tt C+mh,5,12,0;}, it indicates the modification may be any of the stated bases.
The associated confidence values in the {\tt Ml} tag may be used to determine the relative likelihoods between the options.
The example above is equivalent to {\tt C+m,5,12,0;C+h,5,12,0;}, although this will have a different ordering of confidence values in {\tt Ml}.
Note ChEBI codes cannot be used in the multi-modification form (such as the {\tt C+mh} example above).

If the modification is not one of the standard common types (listed below) it can be specified as a numeric ChEBI code.
For example {\tt C+76792,57;} is the same as {\tt C+h,57;}.

An unmodified base of {\tt N} means count any base in {\sf SEQ}, not only those of {\tt N}.
Thus {\tt N+n,100;} means the 101st base is Xanthosine (n), irrespective of the sequence composition.

The standard code types and their associated ChEBI values are listed
below, taken from \emph{Modeling methyl-sensitive transcription factor
motifs with an expanded epigenetic alphabet}, Coby Viner

> et.al. \url{https://www.biorxiv.org/content/10.1101/043794v1}.

Reviewer 3: Minor editorial — these documents do references via footnotes; see e.g. Cock et al in _SAMv1.tex_.

For example ``{\tt Mm:Z:C+m,5,12;C+h,5,12;}'' may have an associated tag of ``{\tt Ml:B:C,204,89,26,130}''.

If the above is rewritten in the multiple-modification form, the probabilities are interleaved in the order presented, giving ``{\tt Mm:Z:C+mh,5,12;  Ml:B:C,204,26,89,130}''.
Note where several possible modifications are presented at the same site, the {\tt Ml} values represent the absolute probabilities of the modification call being correct and not the relative likelihood between the alternatives.
These need not sum to 256 ($p \approx 1.0$) as the remainder represents the probability that none of the modification types are present.
In the example used above, the 6th {\tt C} has 80\% chance of being {\tt 5mC}, 10\% chance of being {\tt 5hmC} and 10\% chance of being an unmodified {\tt C}.

{\tt Ml} values for ambiguity codes give the probability that the modification is one of the possible codes compatible with that ambiguity code.

> For example {\tt Mm:Z:C+C,10; Ml:Z:229} indicates a C call with a probability of 90\% of having some form of unspecified modification.

Reviewer 2: Should this `Ml` tag start with `Ml:Z:C,` as all the others do in the specification? Or is the `C` optional?

Reviewer 4: It should be `Ml:B:C` -- https://github.com/samtools/hts-specs/pull/418/files#diff-e765c6479316309f56b636f88189cdde8c40b854c7bdcce9ee7fe87a4e76febcR596

(meaning a byte array of type `C` -- uint8)

Author: @fbrennen is correct.  The `Z` is clearly an error as the quality values are a byte array.

Color read quality on the original strand of the read. Same encoding as {\sf QUAL}; same length as {\tt CS}.
\end{description}

\subsection{Base modifications}

> Base basications, including base methylation, are represented as a

Reviewer: basications?

Each modified base listed also has a quality value associated with
it.  Given the unmodified base already has a phred likelihood, this
base modification quality should be interpreted as the likelihood of
this modification being correct, rather than the base being unmodified.

\begin{description}
\item[MM:Z:\tagregex{([ACGTN][-+][a-z](,[0-9]+)+;)*}]
\hfill\\
The first character is the unmodified base as seen in the {\sf SEQ}
field, one of {\tt A}, {\tt C}, {\tt G}, {\tt T} or {\tt N}, with
the exception that {\tt N} is used to match any base rather than
strictly {\tt N}.  This is followed by either plus or minus
indicating the strand the modification was observed on (relative to
the recorded strand of {\sf SEQ} with plus meaning same
orientation), and a base modification symbol.  This is then followed
by a comma separated list of how many unmodified seq bases of the
stated base type to skip, stored as a delta to the last and starting
with 0 as the first (or next) base.  Hence this number series is
comparable to the numbers in an {\tt MD} tag.

For example {\tt C+m,5,12,0;} tells us there are three 5-Methylcytosine
bases in the original {\sf SEQ}.  The first 5 {\tt C} bases are

> unmodified and the 6th is modified, as are the 19th (12 inbetween the

Reviewer: `in-between` or simply `bases between` seems better (though my english is from 🇮🇱...so who knows?)

Author: "Between" makes most sense. Thanks.

exist to represent unspecified modifications bases of their respective
canonical base types, plus code {\tt N} to represent an unspecified
modification of any base type.

\begin{center}
\begin{tabular}{lllll}
{\bf Unmodified base} & {\bf Code} & {\bf Abbreviation} & {\bf Name} & {\bf ChEBI} \\
\hline
C & m & 5mC   & 5-Methylcytosine        & 27551 \\
C & h & 5hmC  & 5-Hydroxymethylcytosine & 76792 \\
C & f & 5fC   & 5-Formylcytosine        & 76794 \\
C & c & 5caC  & 5-Carboxylcytosine      & 76793 \\
C & C &       & Ambiguity code; any C mod & \\
\hline
T & g & 5hmU  & 5-Hydroxymethyluracil   & 16964 \\
T & e & 5fU   & 5-Formyluracil          & 80961 \\
T & b & 5caU  & 5-Carboxyluracil        & 17477 \\
T & T &       & Ambiguity code; any T mod & \\
\hline
A & a & 6mA   & 6-Methyladenine         & 28871 \\
A & A &       & Ambiguity code; any A mod & \\
\hline
G & o & 8oxoG & 8-Oxoguanine            & 44605 \\
G & G &       & Ambiguity code; any G mod & \\
\hline
N & n & Xao   & Xanthosine              & 18107 \\
N & N &       & Ambiguity code; any mod & \\
\end{tabular}
\end{center}

\item[MP:Z:\tagvalue{qualities}]
\hfill\\

> The {\tt MP} tag if present lists the Phred qualities of each

Reviewer: ```suggestion
The (optional) {\tt MP} tag lists the Phred qualities of each
```

Author: Agreed, although I removed the parens around optional.

modification listed in the {\tt MM} tag.  The length should match the
number of position deltas from {\tt MM}.  The qualities are encoded in
the same manner as the primary {\sf QUAL} field; one byte per quality
with ASCII value Phred score + 33.  No separators should be present.

> For example {\tt MM:Z:C+m,5,12,3;C+h,57;} may have an associated

Reviewer: would it be better to separate (e.g. with a semi-colon) the qualities that pertain to different modifications? in the example, the qualities would thus become `5EB;/`

Reviewer: (perhaps with a `~` which has been used in the past to separate lists of qualities)

Author: If we have precedence elsewhere then it is good to follow convention.

Reviewing what we have already:
- OX/BZ.  OX is hyphen-concatenated sequence. BZ is defined to be the same length and BZ uses space to separate them (blergh, but non quality value hence chosen).
- RX/QX. As above
- CR/CY. As above
- BC/QT. As above

I've tried hard to expunge that period from my mind, but it seems we got to a consensus somewhere along the line. :-)

So possibly we should do the same here?  It's not quite analogous as we have a comma separated list instead of a sequence, but the same deal applies of choosing an invalid ASCII phred character as a separator.  We can't however claim the strings need to be equal length of course.

Thoughts anyone?  Space at least shows up really well when reading it.

Color read quality on the original strand of the read. Same encoding as {\sf QUAL}; same length as {\tt CS}.
\end{description}

\subsection{Base modifications}

Base modifications, including base methylation, are represented as a
series of edits from the primary unmodified sequence stored in the

> main SAM {\sf SEQ} field.  If the modifed base has no natural

Reviewer: I find this still to be unclear in the presence of alignment...I think that we should store a representation of edits to the originally sequenced (i.e. unaligned, and un-reverse complemented) bases.

Reviewer: ah...I see that the details explains...

unmodified form then this should be stored as ``N''.

Each modified base listed also has a quality value associated with
it.  Given the unmodified base already has a phred likelihood, this
base modification quality should be interpreted as the likelihood of
this modification being correct, rather than the base being unmodified.

\begin{description}
\item[MM:Z:\tagregex{([ACGTN][-+][a-z](,[0-9]+)*;)*}]
\hfill\\

> The first character is the unmodified base as seen in the {\sf SEQ}

Reviewer: "As seen in the SEQ field" is confusing...if the read is rec-comp'ed the first character is the unmodified base s seen in the reverse-complement of SEQ...right?

field, one of {\tt A}, {\tt C}, {\tt G}, {\tt T} or {\tt N}, with
the exception that {\tt N} is used to match any base rather than
strictly {\tt N}.  This is followed by either plus or minus
indicating the strand the modification was observed on (relative to
the original sequenced strand of {\sf SEQ} with plus meaning same
orientation\footnote{Hence a tool that may reverse complement sequences does not need to understand how to manipulate the {\tt MM} and {\tt MP} tags.}), and a base modification symbol.
This is then followed by a comma separated list of how many unmodified seq bases of the
stated base type to skip, stored as a delta to the last and starting
with 0 as the first (or next) base.  Hence this number series is
comparable to the numbers in an {\tt MD} tag.

For example {\tt C+m,5,12,0;} tells us there are three 5-Methylcytosine

> bases in the original {\sf SEQ}.  The first 5 {\tt C} bases are

Reviewer: the "original SEQ" is intending to mean reverse complemented or not depending on bit 0x10, right? I don't think that it's defined anywhere....

unmodified and the 6th is modified, as are the 19th (12 between the
6th and 19th) and 20th.  Similarly {\tt G-m,14;} indicates the 15th
{\tt G} is a 5-Methylcytosine on the opposite strand.

This permits modifications to be listed on either strand with the rare
potential for both strands to have a modification at the same site.

> Note it is permitted for the coordinate list to be empty (for example {\tt MM:Z:C+m;}), which may be used as an explicit indicator that this base modification is not present.

Author: I don't think any can right now, but there were (are?) technologies that unwound the double stranded DNA and sequenced one strand followed by the other strand.   IIRC PacBio SmartBells did this at one point, or maybe still do, and ONT earlier had their "2D" reads which did this.

The point being the format has to cope with what may come along down the road and this is conceivable, albeit a rather unlikely event to need to encode.

> Note it is permitted for the coordinate list to be empty (for example {\tt MM:Z:C+m;}), which may be used as an explicit indicator that this base modification is not present.

Reviewer 5: ONT has had 1D^2 ever since they stopped 2D, from which the forward and reverse-complement reads from a single double-stranded sequence (as separate reads) can be combined into a single sequence. They also plan to re-introduce fully-connected 2D sequences in the future.

If the modification is not one of the standard common types (listed
below) it can be specified as a numeric ChEBI code.  For example
{\tt C+76792,57;} is the same as {\tt C+h,57;}.

An unmodified base of {\tt N} means count any base in {\sf SEQ}, not
only those of {\tt N}.  Thus {\tt N+n,100;} means the 101st base is
Xanthosine (n), irrespective of the sequence composition.

The standard code types and their associated ChEBI values are listed
below, taken from \emph{Modeling methyl-sensitive transcription factor
motifs with an expanded epigenetic alphabet}, Coby Viner
et.al. \url{https://www.biorxiv.org/content/10.1101/043794v1}.
Additionally ambiguity codes {\tt A}, {\tt C}, {\tt G} and {\tt T}
exist to represent unspecified modifications bases of their respective
canonical base types, plus code {\tt N} to represent an unspecified
modification of any base type.

\begin{center}
\begin{tabular}{lllll}
{\bf Unmodified base} & {\bf Code} & {\bf Abbreviation} & {\bf Name} & {\bf ChEBI} \\
\hline
C & m & 5mC   & 5-Methylcytosine        & 27551 \\
C & h & 5hmC  & 5-Hydroxymethylcytosine & 76792 \\
C & f & 5fC   & 5-Formylcytosine        & 76794 \\

> C & c & 5caC  & 5-Carboxylcytosine      & 76793 \\

Reviewer: chebi has this as carboxycytosine (no 'l')

Author: Impressive dedication to proof reading! :-)

I got these terms from the Viner presentation.  Googling shows both are in common usage, but Carboxylcytosine is considerably more common.  I'm unsure if that means the CHEBI database is in error or just using a different accepted spelling.

I'm out of my depth on this one basically.

Reviewer: digging deeper,  it seems that this is the same thing: https://www.ebi.ac.uk/chebi/searchId.do?chebiId=46883

C & C &       & Ambiguity code; any C mod & \\
\hline
T & g & 5hmU  & 5-Hydroxymethyluracil   & 16964 \\
T & e & 5fU   & 5-Formyluracil          & 80961 \\

> T & b & 5caU  & 5-Carboxyluracil        & 17477 \\

Reviewer: probably the same thing (IINAC) but on chebi this is 'uracil-5-carboxylic acid'

Reviewer 6: It's great to see this being incorporated!

I would suggest keeping the "l" for all "carboxyl" prefixes. ChEBI often makes use of names that are uncommon within the applied areas in which the compounds are referenced, as in this case. I've never seen any modified base literature not use the "l".

I've checked the above-mentioned ChEBI IDs and all are correct. You can also find a catalogue of these on our [DNAmod webpage](https://dnamod.hoffmanlab.org), which contains these same recommendations and also lists the ChEBI IDs near the bottom (e.g. [5hmC](https://dnamod.hoffmanlab.org/5-(hydroxymethyl)cytosine.html)).

Hope that helps—really appreciate all the work being done on this!

The optional {\tt MP} tag lists the Phred qualities of each modification listed in the {\tt MM} tag.
The qualities are encoded in the same manner as the primary {\sf QUAL} field; one byte per quality with ASCII value Phred score + 33.
A space character (`{\tt \textvisiblespace}') should be used as a separator between concatenated quality strings when multiple modification lists are present in the {\tt MM} tag.
The length should match the number of position deltas from {\tt MM} plus 1 per space character required.

For example {\tt MM:Z:C+m,5,12,3;C+h,57;} may have an associated
quality tag of {\tt MP:Z:5EB /}.

Quality values for ambiguity codes give the likelihood that the
modification is one of the possible codes compatible with that
ambiguity code.  For example {\tt MM:Z:C+C,10; MP:Z:+} indicates a C
call with an unspecified modification and the phred score of 10 (ASCII
value {\tt +}).  This corresponds to a 90\% chance of the base being
modified.

To represent several possible modifications at the same site the {\tt MP} tag can be used to indicate the probabilities of each possibility.
The values used should be absolute probabilities, not relative between the alternatives.

> For example, a C base that has 90\% chance of being modified with 5mC being three times more likely than 5hmC will encode 5mC with 67.5\% probability ($0.9 * 0.75$)and 5hmC with 22.5\% probability ($0.9 * 0.25$).

Reviewer: If we really want to be able to encode probabilities in the range of 10% - 90% perhaps we should not use Phred which has terrible resolution in that range (10%=10 and 90%=".45"...1?0?)
Please be explicit that the `"` and `&` are encoding phred 1 and 5.

Author: At the time of writing this I hadn't seen any real data.

However I thought about that once I saw how ONT handle things.  They report probabilities as p*255; ie p=0.33 is 84.  They have arrays of values, so if trained on C, 5mC and 5hmC then they'd have 3 score for C which jointly add up to 255 (irrespective of whether they believe the call to be C - this is to be used after combining with the base call probabilities themselves), giving the relative likelihoods of the various modified and unmodified forms.

However it also means they cannot give higher certainties than 99.6%, or around phred 24.  That may well be sufficient.

The question is, do we acknowledge this is different data and is never going to be extremely accurate?   Do we actually care about small fluctuations in probabilities at the low end?  Eg p=0.15 vs p=0.2?

Author: Actually it's simply not possible, which comes down to the phred range again.

If we wish to emit probabilities for more than one possibility, which I believe is how ONT Guppy may go in the future (https://twitter.com/AlbertVilella/status/1151395294879395845), then phred scores will always generate a score of 0, 1 or 2 somewhere.  That's because inherently there will be one call with p < 0.5.

The Illumina answer to this was to do phred+64 and use log-odds instead of log.  Ie log(p/(1-p)) instead of log(1-p).  That gives 0 being p=0.5 and is symmetric so choice A if p=.99 and choice B of p=0.01 gives phred scores of +20 and -20.

However it still begs the question of whether a log scale is appropriate, or whether even knowing the other values helps.  When I was doing consensus calling algorithms I always grumbled that phred isn't good enough.  If you're trying to make a call and you get A with P=0.9 but everything else is C, then it really helps to know whether than remaining 0.1 probabiity on the A was evenly spread across C, G and T or predominantly a remainder for C.  (I just had to go with flat distribution in lieu of anything better.)

I simply don't have enough understanding of the downstream tools to know whether emitting the score for the most likely call is correct (the phred approach), or whether we the benefits of knowing 2nd, 3rd best choices outweighs the extra storage cost.

> This could be represented with {\tt MM:Z:C+m,10;C+h,10; MP:Z:" \&}.

Reviewer: it would be good to quote this or other way separate from the period at the end of the sentence, as it runs together in the PDF (also on line 499 above, though there the semicolon acts as a implicit separator.
